### ISSUE TITLE
Promote final variables that are assigned a non-nullable value

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/language#1023]
With standard promotion, simply checking if a value is null is enough to promote a nullable variable to non-nullable. So this would work: 
```dart
void main() {
  String? temp;
  if (temp != null) {
    print(temp.length);
  }
}
```

Another scenario where a value can never be null is if it is final, and assigned a non-nullable value. For example: 
```dart
String getNonNullString() => "I am not null";

void main() {
  final String? nonNull = getNonNullString();  // not null, and final
  if (nonNull != null) {
    print(nonNull.length);  // works because of promotion
  } 
  print(nonNull.length);  // Error: nonNull is potentially null
}
```
How can `nonNull` ever be null in this case? Another point, usually definite assignment works only in a local scope, but final variables can never be changed, so can't they be promoted no matter where the code accesses it?

**Comment by leafpetersen** [comment]
> How can `nonNull` ever be null in this case?

I'm not entirely sure I follow.  Are you suggesting that we immediately promote based on the return type of `getNonNullString`?  We've considered doing that (there's some discussion of this on an open PR), but then why not just declare it as `final String nonNull;`?  Declaring the variable as nullable seems entirely pointless here.

**Comment by Levi-Lesches** [comment]
Correct, it's completely redundant. I was thinking of it as more of a linter/analyzer thing that can warn the user of this (since I see a lot of new lints for nnbd). I felt that it was kind of important since the compiler actually throws errors for what should be a rather benign case.

**Comment by leafpetersen** [comment]
There are three different proposals linked from the bottom of https://github.com/dart-lang/language/issues/946 .  I'm not 100% sure which one had the discussion around this kind of promotion, but most likely the [last one](https://github.com/dart-lang/language/pull/1009#), which I think is what we're converging on.  Sorry I can't point to the exact comment thread, there's been a lot of discussion so you'll have to wade through it yourself if you're really interested.. :)

**Comment by hacker1024** [comment]
> but then why not just declare it as final String nonNull;?

I have an issue like this where I can't do this because I'm using a package.
#1033
